core/src/main/scala/kafka/coordinator/DelayedJoin.scala

```diff
@@ -31,14 +31,29 @@ import kafka.server.DelayedOperation
  */
 private[coordinator] class DelayedJoin(coordinator: GroupCoordinator,
                                        group: GroupMetadata,
-                                       rebalanceTimeout: Long)
-  extends DelayedOperation(rebalanceTimeout) {
+                                       delayMs: Int,
+                                       val rebalanceTimeOut: Long)
+  extends DelayedOperation(delayMs) {
 
   // overridden since tryComplete already synchronizes on the group. This makes it safe to
   // call purgatory operations while holding the group lock.
   override def safeTryComplete(): Boolean = tryComplete()
 
   override def tryComplete(): Boolean = coordinator.tryCompleteJoin(group, forceComplete)
-  override def onExpiration() = coordinator.onExpireJoin()
-  override def onComplete() = coordinator.onCompleteJoin(group)
+  override def onExpiration() =
+    coordinator.onExpireJoin()
+
+  override def onComplete() = {
+    group synchronized {
```

Comment:
Could this result in a dead-lock? Context is that we have a hierarchy of locks: group lock and purgatory synchronization, in which we would always want to lock on group before iterating on purgatory watchers. That is also why we introduced `safeTryComplete()` (cc @hachikuji ).